Mathlib/FieldTheory/IsSepClosed.lean

```diff
@@ -0,0 +1,174 @@
+  rcases exists_pow_nat_eq x h2 with ⟨z, rfl⟩
+  exact ⟨z, sq z⟩
+
+theorem roots_eq_zero_iff [IsSepClosed k] {p : k[X]} (hsep : p.Separable) :
+    p.roots = 0 ↔ p = Polynomial.C (p.coeff 0) := by
+  refine' ⟨fun h => _, fun hp => by rw [hp, roots_C]⟩
+  cases' le_or_lt (degree p) 0 with hd hd
+  · exact eq_C_of_degree_le_zero hd
+  · obtain ⟨z, hz⟩ := IsSepClosed.exists_root p hd.ne' hsep
+    rw [← mem_roots (ne_zero_of_degree_gt hd), h] at hz
+    simp at hz
+
+theorem exists_eval₂_eq_zero_of_injective {R : Type _} [CommRing R] [IsSepClosed k] (f : R →+* k)
+    (hf : Function.Injective f) (p : R[X]) (hp : p.degree ≠ 0) (hsep : p.Separable) :
+    ∃ x, p.eval₂ f x = 0 :=
+  let ⟨x, hx⟩ := exists_root (p.map f) (by rwa [degree_map_eq_of_injective hf]) (Separable.map hsep)
+  ⟨x, by rwa [eval₂_eq_eval_map, ← IsRoot]⟩
+
+theorem exists_eval₂_eq_zero {R : Type _} [Field R] [IsSepClosed k] (f : R →+* k) (p : R[X])
+    (hp : p.degree ≠ 0) (hsep : p.Separable) : ∃ x, p.eval₂ f x = 0 :=
+  exists_eval₂_eq_zero_of_injective f f.injective p hp hsep
+
+variable (k)
+
+theorem exists_aeval_eq_zero_of_injective {R : Type _} [CommRing R] [IsSepClosed k] [Algebra R k]
+    (hinj : Function.Injective (algebraMap R k)) (p : R[X]) (hp : p.degree ≠ 0)
+    (hsep : p.Separable) : ∃ x : k, aeval x p = 0 :=
+  exists_eval₂_eq_zero_of_injective (algebraMap R k) hinj p hp hsep
+
+theorem exists_aeval_eq_zero {R : Type _} [Field R] [IsSepClosed k] [Algebra R k] (p : R[X])
+    (hp : p.degree ≠ 0) (hsep : p.Separable) : ∃ x : k, aeval x p = 0 :=
+  exists_eval₂_eq_zero (algebraMap R k) p hp hsep
+
+theorem of_exists_root (H : ∀ p : k[X], p.Monic → Irreducible p → Separable p → ∃ x, p.eval x = 0) :
+    IsSepClosed k := by
+  refine ⟨fun p ↦ (fun hsep ↦ Or.inr ?_)⟩
+  intro q hq hdvd
+  simp only [map_id] at hdvd 
+  have hlc : IsUnit (leadingCoeff q)⁻¹ := IsUnit.inv <| Ne.isUnit <|
+    leadingCoeff_ne_zero.2 <| Irreducible.ne_zero hq
+  have hsep' : Separable (q * C (leadingCoeff q)⁻¹) :=
+    Separable.mul (Separable.of_dvd hsep hdvd) ((separable_C _).2 hlc)
+    (by simpa only [← isCoprime_mul_unit_right_right (isUnit_C.2 hlc) q 1, one_mul]
+      using isCoprime_one_right (x := q))
+  have hirr' := hq
+  rw [← irreducible_mul_isUnit (isUnit_C.2 hlc)] at hirr'
+  obtain ⟨x, hx⟩ := H (q * C (leadingCoeff q)⁻¹) (monic_mul_leadingCoeff_inv hq.ne_zero) hirr' hsep'
+  exact degree_mul_leadingCoeff_inv q hq.ne_zero ▸ degree_eq_one_of_irreducible_of_root hirr' hx
+
+theorem degree_eq_one_of_irreducible [IsSepClosed k] {p : k[X]}
+    (hp : Irreducible p) (hsep : p.Separable) : p.degree = 1 :=
+  degree_eq_one_of_irreducible_of_splits hp (IsSepClosed.splits_codomain p hsep)
+
+theorem algebraMap_surjective_of_isIntegral {k K : Type _} [Field k] [Ring K] [IsDomain K]
+    [hk : IsSepClosed k] [Algebra k K] [IsSeparable k K] (hf : Algebra.IsIntegral k K) :
+    Function.Surjective (algebraMap k K) := by
+  refine' fun x => ⟨-(minpoly k x).coeff 0, _⟩
```

Comment:
```suggestion
  refine fun x => ⟨-(minpoly k x).coeff 0, ?_⟩
```
My instinct is to avoid the primed tactics where possible; really I think they were only there to make the port easier (`refine'` is Lean 3 `refine`).